File: src/php/PhantasyIslandPhpRemoteControl/radio/TypeDefView.php
<?php

namespace PhantasyIslandPhpRemoteControl\Radio;

class ViewMaterialChangeCommand
{
    public $materialType = null;
    public $color = null;
    public $opacity = null;
    public $transparent = null;
    public $visible = null;
    public $wireframe = null;
    public $side = null;
    public $depthTest = null;
    public $depthWrite = null;
    public $alphaTest = null;
    public $blending = null;
    public $vertexColors = null;
    public $fog = null;
    public $emissive = null;
    public $emissiveIntensity = null;
    public $metalness = null;
    public $roughness = null;
    public $envMapIntensity = null;
    public $flatShading = null;
    public $shininess = null;
    public $specular = null;
    public $clearcoat = null;
    public $clearcoatRoughness = null;
    public $transmission = null;
    public $ior = null;
    public $thickness = null;
    public $sheen = null;
    public $sheenRoughness = null;
    public $sheenColor = null;
    public $attenuationColor = null;
    public $attenuationDistance = null;
    public $iridescence = null;
    public $iridescenceIOR = null;
    public $iridescenceThicknessRange = null;
    public $specularIntensity = null;
    public $specularColor = null;
    public $reflectivity = null;
    public $dispersion = null;
    public $anisotropy = null;
    public $anisotropyRotation = null;
    public $mapUrl = null;
    public $normalMapUrl = null;
    public $roughnessMapUrl = null;
    public $metalnessMapUrl = null;
    public $emissiveMapUrl = null;
    public $aoMapUrl = null;
    public $alphaMapUrl = null;
    public $bumpMapUrl = null;
    public $displacementMapUrl = null;
    public $normalScale = null;
    public $bumpScale = null;
    public $displacementScale = null;
    public $displacementBias = null;
    public $aoMapIntensity = null;
    public $mapRepeat = null;
    public $mapOffset = null;
    public $mapRotation = null;

    public function to_dict()
    {
        $d = [];
        foreach (get_object_vars($this) as $k => $v) {
            if ($v !== null) $d[$k] = $v;
        }
        return $d;
    }
}

class ViewMaterialChangeCommandSimple
{
    public $color = null;
    public $opacity = null;
    public $transparent = null;
    public $visible = null;
    public $wireframe = null;
    public $side = null;
    public $fog = null;
    public $emissive = null;
    public $emissiveIntensity = null;

    public function to_dict()
    {
        $d = [];
        foreach (get_object_vars($this) as $k => $v) {
            if ($v !== null) $d[$k] = $v;
        }
        return $d;
    }
}

File: src/ex/lib/phantasy_island_remote_control/radio/type_def_view.ex
defmodule PhantasyIslandRemoteControl.Radio.TypeDefView do
  @moduledoc """
  Python 等价类型定义，对应 TypeScript TypeBox 定义的视图材质指令类型。
  """

  defmodule ViewMaterialChangeCommand do
    defstruct [
      :materialType, :color, :opacity, :transparent, :visible, :wireframe, :side,
      :depthTest, :depthWrite, :alphaTest, :blending, :vertexColors, :fog,
      :emissive, :emissiveIntensity, :metalness, :roughness, :envMapIntensity,
      :flatShading, :shininess, :specular, :clearcoat, :clearcoatRoughness,
      :transmission, :ior, :thickness, :sheen, :sheenRoughness, :sheenColor,
      :attenuationColor, :attenuationDistance, :iridescence, :iridescenceIOR,
      :iridescenceThicknessRange, :specularIntensity, :specularColor,
      :reflectivity, :dispersion, :anisotropy, :anisotropyRotation, :mapUrl,
      :normalMapUrl, :roughnessMapUrl, :metalnessMapUrl, :emissiveMapUrl,
      :aoMapUrl, :alphaMapUrl, :bumpMapUrl, :displacementMapUrl, :normalScale,
      :bumpScale, :displacementScale, :displacementBias, :aoMapIntensity,
      :mapRepeat, :mapOffset, :mapRotation
    ]

    def to_map(struct) do
      struct
      |> Map.from_struct()
      |> Enum.reject(fn {_, v} -> is_nil(v) end)
      |> Enum.map(fn {k, v} ->
        case v do
          {a, b} -> {to_string(k), [a, b]}
          _ -> {to_string(k), v}
        end
      end)
      |> Map.new()
    end
  end

  defmodule ViewMaterialChangeCommandSimple do
    defstruct [:color, :opacity, :transparent, :visible, :wireframe, :side, :fog, :emissive, :emissiveIntensity]

    def to_map(struct) do
      struct
      |> Map.from_struct()
      |> Enum.reject(fn {_, v} -> is_nil(v) end)
      |> Enum.map(fn {k, v} -> {to_string(k), v} end)
      |> Map.new()
    end
  end

  def view_material_change_command_from_map(data) do
    tuple_fields = [:iridescenceThicknessRange, :normalScale, :mapRepeat, :mapOffset]
    
    fields = Enum.map(data, fn {k, v} ->
      key = String.to_atom(k)
      if key in tuple_fields and is_list(v) do
        {key, List.to_tuple(v)}
      else
        {key, v}
      end
    end)
    
    struct(ViewMaterialChangeCommand, fields)
  end
end
